### PR TITLE
Week10: Presentation Testing and Refactoring

### PR DESCRIPTION
**Pull request by sunlight0602**
- Use cookies/sessions to transfer error objects
- Acceptance testing
- Refactor views

Run `spec/acceptance_tests` to start acceptance test

### CODE REVIEW

File: Rakefile [hunk 1]
```diff
@@ -24,31 +21,49 @@ task :console do
...
+    sh "RACK_ENV=test bundle exec ruby #{integration_test_path}/gateway_database_spec.rb"
+  end
+
+  desc 'spec checks of View Objects'
+  task :view_objects do
+    sh "RACK_ENV=test bundle exec ruby #{integration_test_path}/view_objects_spec.rb"
+  end
+
+  desc 'run acceptance tests with watir'
+  task :spec_acceptance do
```
Comment (CloudeWu) [inline Rakefile:60]: I think that `spec:acceptance` is already enough. No need for `spec:spec_acceptance`

File: spec/tests/acceptance/acceptance_spec.rb [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+    # When: user wants to send a request
+    # User is able to locate where to input and send query
+    @browser.text_field(name: 'query').present?
+    @browser.button(type: 'submit').present?
+    # Input a valid request
+    @browser.text_field(name: 'query').set(valid_request)
+    @browser.button(type: 'submit').click
+
+    # Then: user jumps to the correct details url
+    @browser.url.include? valid_request.sub! ' ', '+'
```
Comment (CloudeWu) [inline spec/tests/acceptance/acceptance_spec.rb:35]: violate the frozen string literal

File: spec/tests/acceptance/acceptance_spec.rb [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+  after do
+    @browser.close
+    # @headless.destroy
+  end
+
+  job_title = 'frontend engineer'
+  required_skills = %w[TypeScript AWS JavaScript]
+
+  index_url = 'http://127.0.0.1:4001/'
+  details_url = "http://127.0.0.1:4001/details?query=#{job_title.sub! ' ', '+'}"
```
Comment (CloudeWu) [inline spec/tests/acceptance/acceptance_spec.rb:82]: violate the frozen string literal

File: Rakefile [hunk 4]
```diff
@@ -24,31 +21,49 @@ task :console do
...
+    puts 'NOTE: run app in test environment in another process'
+    sh 'ruby spec/tests/acceptance/acceptance_spec.rb'
```
Comment (CloudeWu) [inline Rakefile:62]: You should call the script, rather than running the ruby here

File: spec/skill_spec.rb [hunk 5]
```diff
@@ -1,60 +1,68 @@
 # frozen_string_literal: true
```
Comment (CloudeWu) [inline spec/skill_spec.rb:1]: Sorry. This file should be deleted.

File: spec/tests/acceptance/acceptance_spec.rb (hunk 2 above)
Comment (sunlight0602) [reply spec/tests/acceptance/acceptance_spec.rb]: [Solved](https://github.com/KeroKnows/s-killer/pull/24/commits/230b33619fc6473ac0c9b60689d4a54264614c48)

File: spec/tests/acceptance/acceptance_spec.rb (hunk 3 above)
Comment (sunlight0602) [reply spec/tests/acceptance/acceptance_spec.rb]: [Solved](https://github.com/KeroKnows/s-killer/pull/24/commits/230b33619fc6473ac0c9b60689d4a54264614c48)

File: spec/skill_spec.rb (hunk 5 above)
Comment (sunlight0602) [reply spec/skill_spec.rb]: [Solved](https://github.com/KeroKnows/s-killer/pull/24/commits/c5252bf4e86894cd3e1c80b2c51a0324e7cfe7d0)

File: Rakefile (hunk 4 above)
Comment (sunlight0602) [reply Rakefile]: [Solved](https://github.com/KeroKnows/s-killer/pull/24/commits/4c96709dec9c0a0830c6267c6249883c870b16bd).

But raises an `SignalException: SIGTERM` error after running `rake spec:acceptance` in terminal.
(Tests run just fine, the error is raised after the tests)
I think it is caused by calling the `trap "kill 0" EXIT` multiple times.

File: Rakefile (hunk 1 above)
Comment (sunlight0602) [reply Rakefile]: [Solved](https://github.com/KeroKnows/s-killer/pull/24/commits/4c96709dec9c0a0830c6267c6249883c870b16bd)

File: app/controllers/app.rb [hunk 6]
```diff
@@ -28,76 +32,94 @@ class App < Roda
...
+            end
           end
         end
       end
     end
 
     # request jobs using API if the query has not saved to database
-    class JobCollector
-      def initialize(config)
+    class DataCollector
```
Comment (tienyulin) [inline app/controllers/app.rb:66]: Please move this class to another layer, like domain or infrastructure. It does not belong to controller.

**Review by tienyulin — COMMENTED** [review]
Reviewed

File: app/controllers/app.rb (hunk 6 above)
Comment (sunlight0602) [reply app/controllers/app.rb:66]: Has moved to service objects in this week's progress.

**Pull request merged**